mounts.go

```diff
@@ -0,0 +1,173 @@
+			}
+			targets = append(targets, eim.Target)
+		}
+		runOpts = append(runOpts, mountPropagatorRunOption{&ei})
+	}
+
+	runOpts = append(runOpts, opts...)
+
+	execState := pm.root.Run(runOpts...)
+	pm.root = execState.Root()
+	for _, target := range targets {
+		pm.states[target] = execState.GetMount(target)
+	}
+}
+
+func (pm *persistentMounts) AsRun() llb.RunOption {
+	runOpts := RunOptions{}
+	for _, o := range pm.opts {
+		ei := llb.ExecInfo{}
+		o.SetRunOption(&ei)
+		for i, eim := range ei.Mounts {
+			if st, ok := pm.states[eim.Target]; ok {
+				ei.Mounts[i].Source = st.Output()
+			}
+		}
+		runOpts = append(runOpts, mountPropagatorRunOption{&ei})
+	}
+	return runOpts
+}
+
+func (pm *persistentMounts) Add(opts ...llb.RunOption) {
+	pm.opts = append(pm.opts, opts...)
+}
+
+func (pm persistentMounts) Copy() MountPropagator {
+	newPM := &persistentMounts{
+		root:   pm.root,
+		opts:   make([]llb.RunOption, len(pm.opts)),
+		states: map[string]llb.State{},
+	}
+	copy(newPM.opts, pm.opts)
+	maps.Copy(newPM.states, pm.states)
+	return newPM
+}
+
+func (pm *persistentMounts) ApplyRoot(opts ...llb.StateOption) {
+	for _, opt := range opts {
+		pm.root = opt(pm.root)
+	}
+}
+
+func (pm *persistentMounts) ApplyMount(target string, opts ...llb.StateOption) {
+	mount, ok := pm.states[target]
+	if !ok {
+		mount = llb.Scratch()
+	}
+	for _, opt := range opts {
+		mount = opt(mount)
+	}
+	pm.states[target] = mount
+}
+
+type mountPropagatorRunOption struct {
+	ei *llb.ExecInfo
+}
+
+var _ llb.RunOption = (*mountPropagatorRunOption)(nil)
+
+func (ro mountPropagatorRunOption) SetRunOption(ei *llb.ExecInfo) {
+	ei.Constraints = ro.ei.Constraints
+	if ro.ei.Platform != nil {
+		ei.Platform = ro.ei.Platform
+	}
+	ei.WorkerConstraints = append(ei.WorkerConstraints, ro.ei.WorkerConstraints...)
+	if ro.ei.Metadata.IgnoreCache {
+		ei.Metadata.IgnoreCache = ro.ei.Metadata.IgnoreCache
+	}
+	for k, v := range ro.ei.Metadata.Description {
+		ei.Metadata.Description[k] = v
+	}
+	if ro.ei.Metadata.ExportCache != nil {
+		ei.Metadata.ExportCache = ro.ei.Metadata.ExportCache
+	}
+	for k, v := range ro.ei.Metadata.Caps {
+		ei.Metadata.Caps[k] = v
+	}
+	if ro.ei.Metadata.ProgressGroup != nil {
+		ei.Metadata.ProgressGroup = ro.ei.Metadata.ProgressGroup
+	}
+	if ro.ei.LocalUniqueID != "" {
+		ei.LocalUniqueID = ro.ei.LocalUniqueID
+	}
+	if ro.ei.Caps != nil {
+		ei.Caps = ro.ei.Caps
+	}
+	ei.SourceLocations = append(ei.SourceLocations, ro.ei.SourceLocations...)
+
+	ei.Mounts = append(ei.Mounts, ro.ei.Mounts...)
+	if ro.ei.ReadonlyRootFS {
+		ei.ReadonlyRootFS = ro.ei.ReadonlyRootFS
+	}
+	if ro.ei.ProxyEnv != nil {
+		ei.ProxyEnv = ro.ei.ProxyEnv
+	}
+	ei.Secrets = append(ei.Secrets, ro.ei.Secrets...)
+	ei.SSH = append(ei.SSH, ro.ei.SSH...)
+}
+
+func (pm persistentMounts) Root() llb.State {
+	return pm.root
+}
+
+func (pm persistentMounts) GetMount(target string) llb.State {
+	st, ok := pm.states[target]
+	if !ok {
+		// this should never happen
```

Comment:
Hmm, it's an exported method, so it seems like a caller could easily misuse it. Maybe returning an error is best?

Reply:
Yeah, good point, fixed, not sure what I was thinking there.